sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -652,6 +653,299 @@ case class MapObjects private(
   }
 }
 
+object CollectObjectsToMap {
+  private val curId = new java.util.concurrent.atomic.AtomicInteger()
+
+  /**
+   * Construct an instance of CollectObjects case class.
+   *
+   * @param keyFunction The function applied on the key collection elements.
+   * @param keyInputData An expression that when evaluated returns a key collection object.
+   * @param keyElementType The data type of key elements in the collection.
+   * @param valueFunction The function applied on the value collection elements.
+   * @param valueInputData An expression that when evaluated returns a value collection object.
+   * @param valueElementType The data type of value elements in the collection.
+   * @param collClass The type of the resulting collection.
+   */
+  def apply(
+      keyFunction: Expression => Expression,
+      keyInputData: Expression,
+      keyElementType: DataType,
+      valueFunction: Expression => Expression,
+      valueInputData: Expression,
+      valueElementType: DataType,
+      collClass: Class[_]): CollectObjectsToMap = {
+    val id = curId.getAndIncrement()
+    val keyLoopValue = s"CollectObjectsToMap_keyLoopValue$id"
+    val keyLoopIsNull = s"CollectObjectsToMap_keyLoopIsNull$id"
```

Comment:
tbh I think this new expression should only be used to turn a catalyst map to external map, and we don't need to generalize it. We can even let it only accept a map type input, instead of 2 array inputs.